### PR TITLE
global: pluggable converter

### PR DESCRIPTION
**Pull request by Panos512**
- Adds a pluggable converter to orcid-json.
- Fixes minor bugs.

Signed-off-by: Panos Paparrigopoulos panos.paparrigopoulos@cern.ch

### CODE REVIEW

File: invenio_orcid/tasks.py [hunk 1]
```diff
@@ -24,19 +24,29 @@
 
 from celery import shared_task
 from celery.utils.log import get_task_logger
+
```
Comment (jirikuncar) [inline invenio_orcid/tasks.py]: @Panos512 please use `isort` to organize imports

File: invenio_orcid/tasks.py [hunk 2]
```diff
@@ -24,19 +24,29 @@
...
+from werkzeug.utils import import_string
+
 from .models import ORCIDRecords
 from .proxies import current_orcid
-from .utils import convert_to_orcid, get_authors_credentials
+from .utils import get_authors_credentials
+
 
 logger = get_task_logger(__name__)
+convert_to_orcid = import_string(current_app.config['ORCID_JSON_CONVERTER_MODULE'])
```
Comment (jirikuncar) [inline invenio_orcid/tasks.py]: Never use `current_app` on global level as it's a proxy to application.

File: setup.py [hunk 3]
```diff
@@ -63,7 +63,7 @@
     'invenio-db>=1.0.0a9',
     'invenio-oauthclient>=1.0.0a4',
     'orcid>=0.7.0',
-]
+    ]
```
Comment (jirikuncar) [inline setup.py]: please revert this change

File: invenio_orcid/utils.py [hunk 4]
```diff
@@ -31,3 +31,30 @@ def get_authors_credentials(orcid):
         id=orcid, method='orcid').first()
     user = RemoteAccount.query.filter_by(user_id=raw_user.id_user).first()
     return user.tokens[0].access_token
+
+
+def convert_to_orcid(record):
```
Comment (jirikuncar) [inline invenio_orcid/utils.py:40]: Do we need it right now or can we include proper implementation later?

File: invenio_orcid/config.py [hunk 5]
```diff
@@ -26,6 +26,8 @@
 
 ORCID_SYNCHRONIZATION_ENABLED = True
 
+ORCID_JSON_CONVERTER_MODULE = 'invenio_orcid.utils:convert_to_orcid'
```
Comment (jirikuncar) [inline invenio_orcid/config.py:33]: Can you have a look at Invenio-Records-REST serializers and use compatible API so one can reuse it in more places?
- `ORCID_RECORD_SERIALIZER` (e.g. https://github.com/inveniosoftware/invenio-records-rest/blob/master/invenio_records_rest/config.py#L48)
- API (see https://github.com/inveniosoftware/invenio-records-rest/blob/master/invenio_records_rest/serializers/json.py)

File: invenio_orcid/utils.py (hunk 4 above)
Comment (jirikuncar) [inline invenio_orcid/utils.py:40]: As an example we can use MARC21 schema and `invenio_records.data`.

Comment (Panos512) [inline invenio_orcid/utils.py:40]: I think it's project-specific. 
The goal  of this is givena the record of a paper to produce a json document that orcid will accept.
So that means that there is no generic implementation (as fas as I can think).

Comment (jirikuncar) [inline invenio_orcid/utils.py:40]: > As an example ...

IMHO we don't want to show Zenodo or INSPIRE data model here. MARC21 is neutral in digital library world.

File: invenio_orcid/config.py (hunk 5 above)
Comment (Panos512) [inline invenio_orcid/config.py:33]: do you mean this https://github.com/inspirehep/inspire-next/blob/master/inspirehep/modules/records/serializers/orcid_serializer.py ?

Comment (jirikuncar) [inline invenio_orcid/config.py:33]: @Panos512 yes!

File: invenio_orcid/utils.py (hunk 4 above)
Comment (Panos512) [inline invenio_orcid/utils.py:40]: Oh, now I understand. That's not Inspire's data it's a sample of orcid-json. Actually it's the minimum request orcid will accept.

Comment (jirikuncar) [inline invenio_orcid/utils.py:40]: We should show people how to create a minimal transformation from MARC21 to ORCID model so they can replicate it for their custom data model.

File: invenio_orcid/serializers/orcid_serializer.py [hunk 6]
```diff
@@ -0,0 +1,57 @@
...
+        """Serialize a search result.
+        :param pid_fetcher: Persistent identifier fetcher.
+        :param search_result: Elasticsearch search result.
+        :param links: Dictionary of links to add to response.
+        """
+        records = []
+        for hit in search_result['hits']['hits']:
+            records.append(json.dumps((convert_to_orcid(hit['_source']))))
+
+        return "\n".join(records)
```
Comment (jirikuncar) [inline invenio_orcid/serializers/orcid_serializer.py:68]: Can you use `flask.jsonify` (http://flask.pocoo.org/docs/0.11/api/#flask.json.jsonify)?

File: invenio_orcid/utils.py (hunk 4 above)
Comment (Panos512) [inline invenio_orcid/utils.py:40]: OK. What if I provide a simple example record in MARC with just title and DOI, and provide an implementation (written in pure python) that transform the dictionary representation of MARC21 to the dictionary representation of orcid? (using dojson would be overkill for the sake of an example)

MARC example:

``` xml
<record>
    <datafield tag="024" ind1="7" ind2=" ">
        <subfield code="a">10.1234/foo.bar</subfield>
        <subfield code="2">DOI</subfield>
    </datafield>
    <datafield tag="245" ind1=" " ind2=" ">
        <subfield code="a">The title</subfield>
    </datafield>
</record>
```

Comment (jirikuncar) [inline invenio_orcid/utils.py:40]: Just make sure that you use the dojson serialization of MARC.

```
[
  {
    "other_standard_identifier":[
      {
        "standard_number_or_code":"10.1234/foo.bar",
        "source_of_number_or_code":"DOI"
      }
    ],
    "title_statement":{
      "title":"The title"
    }
  }
]
```
